src/System.Threading.Channels/src/System/Threading/Channels/AsyncOperation.cs

```diff
@@ -273,8 +272,14 @@ public void OnCompleted(Action<object> continuation, object state, short token,
                     ThrowMultipleContinuations();
                 }
 
-                // Queue the continuation.
-                if (sc != null)
+                // Queue the continuation.  We always queue here, even if !RunContinuationsAsynchronously, in order
+                // to avoid stack diving; this path happens in the rare race when we're setting up to await and the
+                // object is completed after the awaiter.IsCompleted but before the awaiter.OnCompleted.
+                if (_schedulingContext == null)
```

Comment:
>if (_schedulingContext == null) [](start = 16, length = 31)

Is it possible at all to have _schedulingContext == null and sc != null? and if yes, shouldn't we still need to do sc.Post at that time?

Reply:
See the code above that assigns _schedulingContext: it's only set to sc if the sc is one we want to respect.

That said, looking at it again there's an existing very-corner-case bug here, pre-existing this change... it's possible sc could be non-null, ts could be non-null, and we want to use the ts, but we'll end up using the sc instead.  I'll fix that separately.  We should be setting sc to null when we go to check for a task scheduler.

Reply:
also it is worth adding some assert here.

---
In reply to: [226717506](https://github.com/dotnet/corefx/pull/32904#discussion_r226717506) [](ancestors = 226717506)